src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNITcpHandle.cs

```diff
@@ -396,217 +366,299 @@ private static IEnumerable<IPAddress> GetHostAddressesSortedByPreference(string
+                            checkWriteLst = new List<Socket>(socketsInFlight.Count);
+                            checkWriteLst.AddRange(socketsInFlight);
+                            checkErrorLst = new List<Socket>(socketsInFlight.Count);
+                            checkErrorLst.AddRange(socketsInFlight);
 
                             SqlClientEventSource.Log.TrySNITraceEvent(nameof(SNITCPHandle), EventType.INFO,
-                                                                      "Determining the status of the socket during the remaining timeout of {0} microseconds.",
-                                                                      socketSelectTimeout);
+                                "Watching pending sockets during the remaining timeout of {0} microseconds.",
+                                socketSelectTimeout);
 
+                            // Socket.Select will return as soon as any socket is readable, writable, or errored
                             Socket.Select(checkReadLst, checkWriteLst, checkErrorLst, socketSelectTimeout);
-                            // nothing selected means timeout
-                        } while (checkReadLst.Count == 0 && checkWriteLst.Count == 0 && checkErrorLst.Count == 0);
+                            // nothing selected means select timed out
+                        } while (checkReadLst.Count == 0 && checkWriteLst.Count == 0 && checkErrorLst.Count == 0 && !timeout.IsExpired);
+                    }
+                    catch (SocketException e)
+                    {
+                        // Socket.Select can throw if one of the sockets has issues. That socket will be in checkErrorLst.
+                        // Log the error and let that socket be removed from socketsInFlight below.
+                        SqlClientEventSource.Log.TryAdvancedTraceEvent(
+                            "{0}.{1}{2}THIS EXCEPTION IS BEING SWALLOWED: {3}", nameof(SNITCPHandle), nameof(ParallelConnect), EventType.ERR, e);
+                        lastError = e;
+                    }
 
-                        // workaround: false positive socket.Connected on linux: https://github.com/dotnet/runtime/issues/55538
-                        isConnected = socket.Connected && checkErrorLst.Count == 0;
+                    if (timeout.IsExpired)
+                    {
+                        SqlClientEventSource.Log.TryAdvancedTraceEvent(
+                            "{0}.{1}{2}ParallelConnect timeout expired.", nameof(SNITCPHandle), nameof(ParallelConnect), EventType.INFO);
+                        break;
                     }
 
-                    if (isConnected)
+                    // See if one of the sockets in the Selected list is connected without errors (exists in read list or write list but not error list)
+                    // Concat read/write lists since we treat them the same from here
+                    checkWriteLst.AddRange(checkReadLst);
```

Comment:
Should we definitely treat the lists of readable and writable sockets in the same way?

The documentation for [Socket.Select](https://learn.microsoft.com/en-us/dotnet/api/system.net.sockets.socket.select?view=net-8.0) says:
> If you are in a listening state, readability means that a call to Accept will succeed without blocking. If you have already accepted the connection, readability means that data is available for reading. In these cases, all receive operations will succeed without blocking. **Readability can also indicate whether the remote Socket has shut down the connection**; in that case a call to Receive will return immediately, with zero bytes returned.

Both handshakes start with a client-initiated write, (the TDS handshake sends a PRELOGIN packet, the TLS handshake sends a Client Hello) so I wouldn't expect the client to be able to read data from the socket immediately. This leaves the scenario I've bolded above - a remote socket accepts the connection, but then immediately shuts it down after the TCP handshake is complete.

I haven't tested this, but it seems that a socket in `checkReadLst` should actually be treated the same way as a socket in `checkErrorLst` - we should skip any sockets contained in either list.

Reply:
I was unsure about this scenario. In the event a server shuts down the socket immediately after the handshake is complete, would this update the Connected property? I think this is the DNN case that is fixed in #2877. Unfortunately, I don't have an environment to test this in. It might be good to write a simple test server to be able to test socket behavior. Anyone have the time / desire and want to drive this forward? If not, I can try to get to it later.

Reply:
I think the test server would be a good idea; it'd need to tackle two situations:
1. Accept, then close a TCP connection (the DNN case, along with Istio/Envoy)
2. Accept a TCP connection, then immediately start sending bytes to it (so SqlClient ignores sockets which don't comply with the TDS spec.)
I'll make a note to pick it up later, once we're happy with SNITcpHandle's behaviour.

Without testing, I'd expect you to be correct about the Connected property; it'd be strange for Socket.Select to reflect a different state to the top-level property on Socket. If we're just interested in testing whether or not the socket is connected, (i.e. #2877, situation 1) I don't think we'd need to include any reference to `checkReadLst` - whether to include it in the list of sockets to check, or to ensure that none of our sockets are in the list.

This just leaves situation 2, where another service is running on that port and immediately starts sending data. We can exclude that from the MultiSubnetFailover socket selection pretty trivially at this stage by making sure that none of the connected sockets are in `checkReadLst` - a server which immediately sends something to every connected client is violating the TDS protocol from the outset.

As a slight aside, while #2877 should fix DNNs, its associated issue opened up an interesting topic: TCP proxies such as Envoy (and by extension, Istio, which was the original service mentioned and which seems to encapsulate Envoy.) These proxies decouple the front-end acceptance of inbound client connections from the establishment of connections to the back-end SQL Servers. I'm fairly sure this wouldn't be compatible with the current MSF implementation, but haven't gotten around to testing that. I'm not actually sure they can/should be made compatible, either: the design of bypassing MSF/read-only routing and putting a TCP-level load balancer in front of X stateful SQL Server containers (which aren't guaranteed to be in sync in the way that read-only routing is supposed to guarantee) strikes me as a bad idea.

Reply:
I couldn't resist trying #1, which led me down a rabbit hole testing with a local mock server. My #2877 fix is unlikely to do anything for the DNN situation. (I added a comment there.) I think we can ignore/discard the checkReadList, as you said.

Need more time to add details here. Will try to add some tomorrow. There are a lot of potential negative scenarios to test here that could lead to a big time suck.

Reply:
Thanks for checking this. The more I think about the edge cases here, the more I think that MultiSubnetFailover should work in terms of application-level connection state. Perhaps we shouldn't just pick the first socket to establish a TCP connection, but should pick the first IP address to respond to a PRELOGIN request. That eliminates the need for the pause to allow connections to stabilise and lets us fold the various failure states into a single code path, but brings a host of other implications which need a lot more testing.

Reply:
After validating, I changed the logic to only consider the write list. As for the DNN issue, after talking with CSS, it seems their scenario is very specific. They are trying to replicate DNNs on-prem and their networking stack is behaving differently than Azure setups. This is unlikely to be common to anyone else so I'm leaving out any changes specific to that issue.